### PR TITLE
Proposal 1

### PR DESCRIPTION
**Pull request by paulofaria**
**This should not be considered a proposal**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,544 @@
...
+
+```swift
+public struct Data {
+    public var bytes: [Byte]
+}
+```
+
+### Motivation
+
+Having a type for binary data makes low level operations a lot easier. Specially when interfacing with `C`.
```
Comment (tannernelson) [inline README.md]: I would avoid using language like `a lot` or even `easier`. Technical documentation should be technical. 
> Binary data types facilitate standardization of low level operations.

File: README.md [hunk 2]
```diff
@@ -1 +1,544 @@
...
+### Motivation
+
+Having `URI` as a `struct` allows it to conform to protocols in extensions.
+
+## HTTPRequest
+
+The `HTTPRequest` type represents an HTTP request.
+
+```swift
+public struct HTTPRequest: HTTPMessage {
```
Comment (tannernelson) [inline README.md]: Why is this a `struct`?

File: Sources/S4.swift [hunk 3]
```diff
@@ -0,0 +1,314 @@
+public typealias Byte = UInt8
```
Comment (tannernelson) [inline Sources/S4.swift]: We should split these up into individual files. 

`HTTPRequest.swift`, `HTTPResponse.swift`, etc.

File: README.md (hunk 1 above)
Comment (paulofaria) [inline README.md]: haha what would be technical term for `lot` and `easier`?

Comment (tannernelson) [inline README.md]: ? `facilitate`. I included a sample replacement in my comment. :P

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: It says on the motivation. It's better to have structs than protocols sometimes. cause with structs you can make it adopt other protocols in an extension and protocols don't allow that.

File: Sources/S4.swift (hunk 3 above)
Comment (paulofaria) [inline Sources/S4.swift]: OK we can do that later, otherwise we're gonna lose current comments.

File: README.md (hunk 1 above)
Comment (tannernelson) [inline README.md]: Writing technical prose is a whole profession. There is so much important stuff there. I think we should treat S4 like the ISO would treat a standardization. We should look to something like the ANSI C spec and strive to make it as concise, clear, and technical.

File: README.md (hunk 2 above)
Comment (tannernelson) [inline README.md]: This will require a conversion to and from a struct for every request / response in any `S4` app. Complying with S4 should have no inherent performance affects on the application if possible.

**Comment by tannernelson** [comment]
I see two different ways the `HTTPRequest` could work. 

```swift
import S4

class MyCoolRequest: S4.HTTPRequest {
	
}
```


``swift
import S4

extension MyCoolRequest: S4.HTTPRequestConvertible {
	var s4HTTPRequest: S4.HTTPRequest {
		return S4.HTTPRequest(self.method, self.....)
	}
}
```

File: README.md (hunk 1 above)
Comment (paulofaria) [inline README.md]: Yeah, I'm not a native english speaker so feel free to rewrite all my sentences hahaha.

**Comment by paulofaria** [comment]
Out of curiosity, why use a `class` for the request?

**Comment by bradhilton** [comment]
First off, as I read this proposal I'm reminded that `enum`s are impossible (difficult?) to represent as a protocol, so I think it make sense that anything that _must_ be represented as an `enum` be a concrete type as defined in the proposal. Based on our discussion I think that `HTTPMethod` and `HTTPStatus` are well defined.

**Comment by paulofaria** [comment]
About the enums. Protocols can't be used as keys to dictionaries and could be desirable for `HTTPMethod` specially.

**Comment by bradhilton** [comment]
I guess I should mention that `String` or `Int` could be used as keys without the enums...

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: I can't see why making `HTTPRequest` a struct could be a problem. It has everything you need for custom computed properties on extensions. Specially cause it has `storage` which means you can add "stored properties" on extensions. 

Example:

````swift
extension HTTPRequest {
    public var pathParameters: [String: String] {
        set {
            storage["pathParameters"] = newValue
        }
        get {
            return storage["pathParameters"] as? [String: String] ?? [:]
        }
    }
}
```

So you don't need to convert it. You can just add more functionality through extensions.

**Comment by tannernelson** [comment]
@paulofaria Vapor uses a class for Request, and I could see other frameworks wanting to do that. Vapor uses a lot of callbacks and copying the Request struct for each of those seems like a lot of extra CPU work.

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: > @paulofaria Vapor uses a class for Request, and I could see other frameworks wanting to do that. Vapor uses a lot of callbacks and copying the Request struct for each of those seems like a lot of extra CPU work. Forcing a struct for the Request data structure in an app seems like too much of an opinion.

`structs` are copy on write so there wouldn't be a performance hit.

Comment (tannernelson) [inline README.md]: Still seems like too much opinion to me. We are just trying to make projects work together, not show them how to work.

Comment (paulofaria) [inline README.md]: https://www.youtube.com/watch?v=av4i3x-aZbM

Comment (tannernelson) [inline README.md]: Not saying one is right or wrong, just saying that could be a detractor for people looking to make their project comply with S4. We should start with the absolute minimum here.

Comment (tannernelson) [inline README.md]: But I will watch that video later. It looks interesting. :)

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.DS_Store
```
Comment (LoganWright) [inline .gitignore]: We should make a `.gitignore` w/ at least this in it for now

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: > Still seems like too much opinion to me. We are just trying to make projects work together, not show them how to work.

I don't agree completely. I think standards should have an opinion if it's proven to be the best. I'm not saying **this** is the best opinion. I'm saying we should look for one and commit to it. As I said before, standards and good practice goes together. At least from my point of view `Swift X` wouldn't be only about making projects work together. It would be to make projects work together using the proven best practices.

File: .gitignore (hunk 4 above)
Comment (paulofaria) [inline .gitignore]: but this is a `.gitignore`. 🤔

Comment (tannernelson) [inline .gitignore]: I think he meant on master.

**Comment by tannernelson** [comment]
@paulofaria So are you saying Requests in a project conforming to `S4` should be sub-structs of `S4.HTTPRequest`?

**Comment by paulofaria** [comment]
We should comment on the diff, not on the PR itself. this way we can have a logical sequence of messages. So I'm gonna reply in the HTTPRequest section.

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: > @paulofaria So are you saying Requests in a project conforming to S4 should be sub-structs of S4.HTTPRequest?

```swift
import S4

struct MyCoolRequest: S4.HTTPRequest {
    //inherits all the properties from S4.HTTPRequest
}
```

No. I don't think that's even possible. The idea would be to simply add extensions. For example. I see Vapor uses `public let cookies: [String: String]`. So you could add it in an extension like this.

```swift
import S4

// You can even typealias to a name you prefer
typealias MyCoolRequest = HTTPRequest

extension HTTPRequest {
    public var cookies: [String: String] {
        set {
            // transform the dictionary in header values and add them to headers property
        }
        get {
            // retrieve cookies from the headers property
        }
    }
}
```

Comment (tannernelson) [inline README.md]: I see. So in essence the classes in every project are based off of the `S4` structs. This seems a lot more restricting than a protocol would be. What were the cons to using a protocol here?
@LoganWright

Comment (tannernelson) [inline README.md]: > For example you could make it Hashable (in case we don't make the HTTPRequest protocol inherit from Hashable in its definition).

I don't see any problem with explicitly declaring all conforming protocols in the original definition. If that's the only caveat.

**Comment by tannernelson** [comment]
@paulofaria Good idea, I'll stop spamming the general PR comments. >_<

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: The problem is that we would have to define beforehand all protocols, and that's what's limiting imo. the only limitation about having request be a `struct` would be that it wouldn't be a `class`. but it could be a `class` and then the limitation would be that it is not a `struct`. either way I think it would be less limiting if it is a concrete type instead of a protocol. Personally I prefer structs because of all the advantages we know value types have over reference types. But that's up for discussion as everything here.

Comment (paulofaria) [inline README.md]: @tannernelson you can't run away from opinion.

Request as protocol:
- S4 would define the protocols request inherits from (opinionated)
- The concrete implementation could be struct or class

Request as concrete type:
- request could inherit from whatever protocol it wants
- request would have to be a class or a struct (opinionated)

Comment (paulofaria) [inline README.md]: Zewo's request is a struct and conforms to:

`CustomStringConvertible`
`CustomDebugStringConvertible`
`Hashable`

Vapor's request is a class and doesn't conform to any protocol

Inquiline (Nest's) is a struct and conforms to:

`CustomStringConvertible`
`CustomDebugStringConvertible`

Kitura's request is a class and conforms to:

`BlueSocketWriter`
`CurlInvokerDelegate`

Comment (bradhilton) [inline README.md]: I think for most cases value semantics are better than reference semantics. I think that classes are best for application state (which I think we should avoid introducing in S4) and perhaps for edge cases where performance may actually be important (see discussion on `Data`). `Request` is not large or representing application state, so I believe it is best left as a struct.

Comment (tannernelson) [inline README.md]: I'm confused about this part. If `Vapor.Request` conforms to `S4.HTTPRequestProtocol` I can still make `Vapor.Request` conform to other protocols by extending it because it's a concrete class.

File: README.md [hunk 5]
```diff
@@ -1 +1,558 @@
...
+    public var path: String?
+    public var query: [String: String]
+    public var fragment: String?
+}
+```
+
+### Motivation
+
+Having `URI` as a `struct` allows it to conform to protocols in extensions.
+
```
Comment (bradhilton) [inline README.md]: I like the URI definition. Only comment is that `query` can't be a dictionary as it is possible to have multiple of the same query param (Kinda like headers). Also I believe it's possible to have a query param without a value. So it might be better as `public var query: [(String: String?)]`.

Comment (tannernelson) [inline README.md]: A query param with a value `&no_value=&value=1` could be argued as just being an empty string.

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: Request being `Hashable` is not that important for Zewo, so if request is a protocol with at least:

`CustomStringConvertible`
`CustomDebugStringConvertible`

Zewo would be ok
Vapor would have to implement them (trivial)
Inquiline would be ok
Kitura would simply not work.

If request is a `struct`:

Zewo would be ok
Vapor would have to change it to a struct (are you really using reference type semantics or is this just for perfomance?)
Inquiline would be ok
Kitura could work or not depending if they use reference type semantics

If request is a `class`:

Zewo would have to change to a class
Vapor would be ok
Inquiline would have to change to a class
Kitura would be ok

File: .gitignore (hunk 4 above)
Comment (LoganWright) [inline .gitignore]: lol, my bad! It's been a long week!

File: README.md (hunk 2 above)
Comment (tannernelson) [inline README.md]: Why couldn't you just have protocol extensions that conform to `CustomStringConvertible`, etc. Then the frameworks wouldn't have to implement that?

Also what about Kitura would "simply not work"?

Comment (paulofaria) [inline README.md]: > I'm confused about this part. If Vapor.Request conforms to S4.HTTPRequestProtocol I can still make Vapor.Request conform to other protocols by extending it because it's a concrete class.

Yes you can, but all other protocols in `S4` take `S4.HTTPRequest` request not `MyCoolRequest`. So only the properties defined in `S4.HTTPRequest` (by definition or extension) would be useful across different projects.

File: README.md [hunk 6]
```diff
@@ -1 +1,558 @@
...
+```swift
+let bytesWritten = data.withUnsafeBufferPointer {
+    write(fd, $0.baseAddress, $0.count)
+}
+```
+
+`Data` could be simply a typealias:
+
+```swift
+typealias Data = [Byte]
```
Comment (LoganWright) [inline README.md]: `public`

File: README.md [hunk 7]
```diff
@@ -1 +1,558 @@
...
+
+```swift
+public typealias Byte = UInt8
+```
+
+### Motivation
+
+`Byte` is clearer than `UInt8` in the context of binary data.
+
+## Data
```
Comment (LoganWright) [inline README.md]: I think we might want to remove this section in the interim since there's a lot of opinions, it likely warrants its own thread.

Also conflict w/ `typealias Data = [Byte]`

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: > Why couldn't you just have protocol extensions that conform to CustomStringConvertible, etc. Then the frameworks wouldn't have to implement that?

Yeah we could have that. And frameworks could extend them if they want.

> Also what about Kitura would "simply not work"?

This is just conjecture, but I think It wouldn't work because the request depends on:

`BlueSocketWriter`
`CurlInvokerDelegate`

If request is a protocol, kitura's request wouldn't be able to conform to that protocol. And it seems that these protocols are deep rooted in kitura's architecture. But maybe it could be possible to get rid of them or wrap a request that conforms to `S4.HTTPRequest` inside kitura's request. It surely wouldn't be as easy as in the other frameworks.

Comment (tannernelson) [inline README.md]: Why wouldn't Kitura's request be able to conform to S4.RequestProtocol if it depends on those?

Comment (paulofaria) [inline README.md]: > Why wouldn't Kitura's request be able to conform to S4.RequestProtocol if it depends on those?

It could depend on them but only internally. As soon as the request is passed to another module, like the server for example. it will receive `S4.HTTPRequest`, which won't depend on, `BlueSocketWriter` and `CurlInvokerDelegate. Cause `KituraRequest` (fictional) is the one that depends on them, not  `S4.HTTPRequest`.

Comment (paulofaria) [inline README.md]: If request is a protocol: 

```swift
import S4

class MyCoolRequest: HTTPRequest {
    // standard
    public var method: HTTPMethod
    public var uri: URI
    public var version: HTTPVersion
    public var headers: HTTPHeaders
    public var body: HTTPBody
    public var storage: Storage = [:]  

    // this is useless
    public var cookies: [String: String]
}

extension MyCoolRequest {
    // this is useless
    public var contentLength: Int { ... }
}

extension HTTPRequest: Hashable { ... } // Not Possible

extension HTTPRequest {
    // this will be visible in the framework cause it extends HTTPRequest
    public var contentType: MediaType { ... }
}
```

Comment (paulofaria) [inline README.md]: I'm considering that `request` in this code is `S4.HTTPRequest` (you can typealias it to `MyCoolRequest` if you want to, but it would only change the name)

```
app.get("welcome") { request in
    return "Hello"
}
```

So only properties defined in  `S4.HTTPRequest` would be visible.

Comment (tannernelson) [inline README.md]: Kitura would be passing KituraRequest internally if both modules relied on their proprietary protocols. 

Also, if you passed a KituraRequest through as a S4.RequestProtocol, couldn't you cast it back to a KituraRequest?

Comment (paulofaria) [inline README.md]: The other option would be to construct a `MyCoolVaporRequest` from a `S4.HTTPRequest` and then `request` would truly be `MyCoolVaporRequest`. Then you would have to convert the response to `S4.HTTPResponse` (in case it isn't). 

```swift
app.get("welcome") { request in
    return "Hello"
}
```

Comment (tannernelson) [inline README.md]:
```swift
app.get("welcome") { (request: Vapor.Request) in
    return "Hello"
}
```

That should definitely be a Vapor request there. Otherwise the closure won't have access to any proprietary goodies vapor offers.

Comment (paulofaria) [inline README.md]: And I just realized another problem. I'm thinking we have a strong reason to settle for a `struct` or a `class`. cause if some other module that conforms to `S4` expects a value type and gets a reference type, things could go really bad.

Comment (tannernelson) [inline README.md]: I think I'm finally seeing the difference.

The way I imagine it working is that when Vapor receives an `S4.Request` it converts it to a Vapor request. Where as you (I think?) are saying that Vapor would not need to convert it, it would simply add all it's proprietary goodies as extensions to `S4.Request`

Comment (paulofaria) [inline README.md]: > That should definitely be a Vapor request there. Otherwise the closure won't have access to any proprietary goodies vapor offers.

That's what I'm saying. those goodies should be in `S4.HTTPRequest` extensions if you get a `S4.HTTPRequest` (which could be typaliased to `VaporRequest) **or** you would have to construct `VaporRequest` from `S4.HTTPRequest` and them `VaporResponse`  to `S4.HTTPResponse`.

Comment (paulofaria) [inline README.md]: And you haven't answered if the only reason vapor's request is a class is for performance. Do you guys use reference type semantics for requests in vapor?

Comment (paulofaria) [inline README.md]: > The way I imagine it working is that when Vapor receives an S4.Request it converts it to a Vapor request. Where as you (I think?) are saying that Vapor would not need to convert it, it would simply add all it's proprietary goodies as extensions to S4.Request

Exactly, that would be better than converting, cause you would get that overhead (but not that much cause you would have to convert in the extensions anyway). So it would be pretty much the same.

We could even have some of these goodies in a module outside of Vapor or Zewo and then Zewo and Vapor could use them if they want, or any other middleware or framework, etc.

**Comment by LoganWright** [comment]
I would like to propose one final time, and then I will resist, but I think we will be significantly more productive if we break this up.

The contents included are far to comprehensive to appropriately discuss in the message board style. Each of the components would benefit from a unique pull request that has a good focus.

Once we agree on these, one by one, they can be merged into the repo. Right now, it's way too much to follow and we can't merge any of it until we agree on all of it.

@tannernelson @paulofaria @bradhilton

File: README.md (hunk 2 above)
Comment (tannernelson) [inline README.md]: Okay this makes sense. For the Vapor drivers I was converting values and that worked well. It allows you to keep all conformance code in one file which could be a huge gain. With your method, it involves changes that touch more areas, but could result in less implementation code.

I think we should make this difference clear and see what everyone's say is.

And in terms of why Vapor is using classes, no thought has been put into that. It was my understanding that structs were copied when passed into a function, so I've never really used them. Theoretically it seems to me like there should only be one copy of the Request... but like I said, i still need to wrap my head around the way Swift deals with structs. I just assume they function like C structs.

File: README.md (hunk 5 above)
Comment (paulofaria) [inline README.md]: Yeah. I think an empty string would work.

**Comment by tannernelson** [comment]
@LoganWright I agree 100%. We might as well split them up now. It's probably not going to get easier.

File: README.md (hunk 5 above)
Comment (bradhilton) [inline README.md]: I think the point is that you can a param that works as a flag e.g. `?flag&param=value`.
I think the value should be an optional `String`. We might also want to either have a typealias, protocol or struct for `QueryItem` that represents the name and value of a query argument.

Comment (tannernelson) [inline README.md]: `?flag&` makes sense as optional
but `?flag=&` should be an empty string.

File: README.md (hunk 2 above)
Comment (paulofaria) [inline README.md]: They're copy-on-write, which means they're only copied when something inside the `struct` changes. It's like a reference type that creates a copy automatically every time something changes. The copy is fast cause structs memory layouts are fixed so the copy can be done in `O(1)`. Overall `structs`s are fast and safer. You should only use reference types when you need reference type semantics. If you're using a `class` like a `struct` it should be a `struct`.

Comment (tannernelson) [inline README.md]: Can you give an example of a "reference type semantic"?

Comment (paulofaria) [inline README.md]: If you need inheritance or if you want to change a value in the class and let it be reflected elsewhere where another type is holding a reference to it. none of them are needed for a request. a request is a plain old data structure and that's what `structs` and value types are for.

File: README.md (hunk 5 above)
Comment (paulofaria) [inline README.md]: can we move this into a separate PR?
